// ngrx-forms/esm2015/src/status-css-classes.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Lists the available status class names based on the property
 * they are depending on.
 */
export const NGRX_STATUS_CLASS_NAMES = {
    isValid: 'ngrx-forms-valid',
    isInvalid: 'ngrx-forms-invalid',
    isDirty: 'ngrx-forms-dirty',
    isPristine: 'ngrx-forms-pristine',
    isTouched: 'ngrx-forms-touched',
    isUntouched: 'ngrx-forms-untouched',
    isSubmitted: 'ngrx-forms-submitted',
    isUnsubmitted: 'ngrx-forms-unsubmitted',
    isValidationPending: 'ngrx-forms-validation-pending',
};
export class NgrxStatusCssClassesDirective {
    set ngrxFormControlState(state) {
        this.state = state;
    }
    set ngrxFormState(state) {
        this.state = state;
    }
    get isValid() {
        return this.state.isValid;
    }
    get isInvalid() {
        return this.state.isInvalid;
    }
    get isDirty() {
        return this.state.isDirty;
    }
    get isPristine() {
        return this.state.isPristine;
    }
    get isTouched() {
        return this.state.isTouched;
    }
    get isUntouched() {
        return this.state.isUntouched;
    }
    get isSubmitted() {
        return this.state.isSubmitted;
    }
    get isUnsubmitted() {
        return this.state.isUnsubmitted;
    }
    get isValidationPending() {
        return this.state.isValidationPending;
    }
}
NgrxStatusCssClassesDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: NgrxStatusCssClassesDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NgrxStatusCssClassesDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.13", type: NgrxStatusCssClassesDirective, selector: "form[ngrxFormState],[ngrxFormControlState]", inputs: { ngrxFormControlState: "ngrxFormControlState", ngrxFormState: "ngrxFormState" }, host: { properties: { "class.ngrx-forms-valid": "this.isValid", "class.ngrx-forms-invalid": "this.isInvalid", "class.ngrx-forms-dirty": "this.isDirty", "class.ngrx-forms-pristine": "this.isPristine", "class.ngrx-forms-touched": "this.isTouched", "class.ngrx-forms-untouched": "this.isUntouched", "class.ngrx-forms-submitted": "this.isSubmitted", "class.ngrx-forms-unsubmitted": "this.isUnsubmitted", "class.ngrx-forms-validation-pending": "this.isValidationPending" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: NgrxStatusCssClassesDirective, decorators: [{
            type: Directive,
            args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'form[ngrxFormState],[ngrxFormControlState]',
                }]
        }], propDecorators: { ngrxFormControlState: [{
                type: Input
            }], ngrxFormState: [{
                type: Input
            }], isValid: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isValid}`]
            }], isInvalid: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isInvalid}`]
            }], isDirty: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isDirty}`]
            }], isPristine: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isPristine}`]
            }], isTouched: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isTouched}`]
            }], isUntouched: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isUntouched}`]
            }], isSubmitted: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isSubmitted}`]
            }], isUnsubmitted: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isUnsubmitted}`]
            }], isValidationPending: [{
                type: HostBinding,
                args: [`class.${NGRX_STATUS_CLASS_NAMES.isValidationPending}`]
            }] } });
