// ngrx-forms/esm2015/src/group/local-state-directive.js
import { Directive, EventEmitter, Output } from '@angular/core';
import { NgrxFormDirective } from './directive';
import * as i0 from "@angular/core";
export class NgrxLocalFormDirective extends NgrxFormDirective {
    constructor() {
        super(null);
        this.ngrxFormsAction = new EventEmitter();
    }
    dispatchAction(action) {
        this.ngrxFormsAction.emit(action);
    }
}
NgrxLocalFormDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: NgrxLocalFormDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NgrxLocalFormDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.13", type: NgrxLocalFormDirective, selector: "form[ngrxFormState][ngrxFormsAction]", outputs: { ngrxFormsAction: "ngrxFormsAction" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: NgrxLocalFormDirective, decorators: [{
            type: Directive,
            args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'form[ngrxFormState][ngrxFormsAction]',
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { ngrxFormsAction: [{
                type: Output
            }] } });
